### PR TITLE
feat(s3-asset): add httpUrl and s3ObjectUrl

### LINKED ISSUE #7221: s3 asset object s3 url
add a method or property to get the 's3://bucket/key' style url from the Asset in aws-s3-assets

### Use Case

In many places s3 assets are used in other constructs. In many projects I see:
people doing something like: `s3://${myAsset.s3BucketName}/${myAsset.s3ObjectKey}`

you could easily make small mistakes doing this, while it's such a common need.

### Proposed Solution

add a property to Asset s3ObjectUrl which holds `s3://${myAsset.s3BucketName}/${myAsset.s3ObjectKey}`

### Other

* [x] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by christophgysin**
### Commit Message
feat(s3-asset): add httpUrl/s3ObjectUrl (#7509)

fixes #7221
### End Commit Message

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/core/lib/stack.ts [hunk 1]
```diff
@@ -552,9 +552,10 @@ export class Stack extends Construct implements ITaggable {
     const s3Filename = Fn.select(1, Fn.split(cxapi.ASSET_PREFIX_SEPARATOR, encodedKey));
     const objectKey = `${s3Prefix}${s3Filename}`;
 
-    const s3Url = `https://s3.${this.region}.${this.urlSuffix}/${bucketName}/${objectKey}`;
+    const httpUrl = `https://s3.${this.region}.${this.urlSuffix}/${bucketName}/${objectKey}`;
+    const s3ObjectUrl = `s3://${bucketName}/${objectKey}`;
```
Comment (iliapolo) [inline packages/@aws-cdk/core/lib/stack.ts:556]: Can we add a unit test for this? I see there is an integration test in the `s3-assets` module that indirectly checks this, but i'd prefer a more explicit test in the `core` module as well.

File: packages/@aws-cdk/aws-s3-assets/README.md [hunk 2]
```diff
@@ -34,7 +34,7 @@ to an S3 bucket during deployment.
 
  * `s3BucketName` - the name of the assets S3 bucket.
  * `s3ObjectKey` - the S3 object key of the asset file (whether it's a file or a zip archive)
- * `s3Url` - the S3 URL of the asset (i.e. https://s3.us-east-1.amazonaws.com/mybucket/mykey.zip)
+ * `httpUrl` - the S3 HTTP URL of the asset (i.e. https://s3.us-east-1.amazonaws.com/mybucket/mykey.zip)
```
Comment (iliapolo) [inline packages/@aws-cdk/aws-s3-assets/README.md:37]: Lets add `s3ObjectUrl` here as well :)

Comment (christophgysin) [reply packages/@aws-cdk/aws-s3-assets/README.md:37]: Added.

File: packages/@aws-cdk/core/lib/stack.ts (hunk 1 above)
Comment (christophgysin) [reply packages/@aws-cdk/core/lib/stack.ts:556]: Sure!

I'm not quite sure how to set up the expectation though. The returned assetLocation contains tokens:
```ts
const assetLocation = stack.addFileAsset(...);
...
test.deepEqual(assetLocation, {
  bucketName: '${Token[TOKEN.79]}',
  objectKey: '${Token[TOKEN.82]}${Token[TOKEN.84]}',
  httpUrl: 'https://s3.${Token[AWS::Region.4]}.${Token[AWS::URLSuffix.1]}/${Token[TOKEN.79]}/${Token[TOKEN.82]}${Token[TOKEN.84]}',
  s3ObjectUrl: 's3://${Token[TOKEN.79]}/${Token[TOKEN.82]}${Token[TOKEN.84]}',
  s3Url: 'https://s3.${Token[AWS::Region.4]}.${Token[AWS::URLSuffix.1]}/${Token[TOKEN.79]}/${Token[TOKEN.82]}${Token[TOKEN.84]}',
});
```
Is there a better way to test this?

Comment (christophgysin) [reply packages/@aws-cdk/core/lib/stack.ts:556]: @iliapolo Any feedback on the above? Does a test like this make sense? Is there a better way to test this?

**Comment by christophgysin** [comment]
Added the test as suggested. Another option would be something like this:
```ts
const {
  region,
  urlSuffix,
} = stack;

const escape = (str: string) => str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
const bucketName = escape(assetLocation.bucketName);
const objectKey = escape(assetLocation.objectKey);

const {
  httpUrl,
  s3ObjectUrl,
  s3Url,
} = assetLocation;

test.ok(RegExp(`^https://s3\.${region}\.${urlSuffix}/${bucketName}/${objectKey}$`).test(httpUrl));
test.ok(RegExp(`^s3://${bucketName}/${objectKey}$`).test(s3ObjectUrl));
test.equal(s3Url, httpUrl); // for backwards compatibility
```

**Comment by iliapolo** [comment]
@christophgysin Looks like we (I) waited to long...and now there is a conflict. Can you have a look? Looks great otherwise 👍

**Comment by christophgysin** [comment]
@iliapolo I will update the PR

**Comment by christophgysin** [comment]
@eladb It seems that the Semantic Pull Request job is stuck? Is there anything I can do to move this forward?

**Pull request merged**